### PR TITLE
Shriya/tp overlap patch

### PR DESCRIPTION
**Pull request by ShriyaPalsamudram**
Upstreaming @erhoo82's branch: [tp_overlap_patch](https://github.com/erhoo82/TransformerEngine/commits/tp_overlap_patch)

### CODE REVIEW

File: transformer_engine/pytorch/cpp_extensions.py [hunk 1]
```diff
@@ -756,6 +756,7 @@ def fp8_gemm(
     ub_algo: tex.UbufOverlapAlgo = None,
     ub: Union[tex.UbufCommOverlap, tex.UbufP2PCommOverlap] = None,
     extra_output_tensor: torch.Tensor = None,
+    layer_name: int = 9,
```
Comment (ptrendx) [inline transformer_engine/pytorch/cpp_extensions.py:759]: Why 9?

**Comment by ShriyaPalsamudram** [comment]
@ptrendx @ksivaman please review, thanks.

File: transformer_engine/pytorch/csrc/userbuffers/userbuffers.h [hunk 2]
```diff
@@ -216,7 +217,7 @@ void userbuffers_recv(const int srchandler, const size_t srcoffset, const int ds
 void userbuffers_alltoall_send(const int srchandler, const size_t srcoffset, const int dsthandler,
                                const size_t dstoffset, const size_t bytes, communicator *comm,
                                cudaStream_t stream = 0);
-void userbuffers_alltoall_recv(communicator *comm, cudaStream_t stream = 0);
+//void userbuffers_alltoall_recv(communicator *comm, cudaStream_t stream = 0);
```
Comment (ptrendx) [inline transformer_engine/pytorch/csrc/userbuffers/userbuffers.h:220]: Same

File: transformer_engine/pytorch/module.py [hunk 3]
```diff
@@ -2880,6 +2887,7 @@ def backward(
                     gelu_input=fc1_out,
                     ub_algo=tex.UbufOverlapAlgo.SPLIT_PIPELINED_AG if ctx.ub_split_ag else None,
                     ub=ctx.ub_obj_gradout if ctx.ub_split_ag else None,
+                    layer_name=4,
```
Comment (ksivaman) [inline transformer_engine/pytorch/module.py:2890]: Why are all these hardcoded? @ShriyaPalsamudram

**Review by timmoon10 — CHANGES_REQUESTED** [review]
If I understand correctly, the only functional change is to change some memory accesses to use `atomicAdd_system`. We also add some more information to an error message, which requires plumbing in the `layer_name` throughout the whole PyTorch infrastructure. If this debugging info is not absolutely essential, I feel the extra messiness is not worth it. At the very least, `layer_name` is a really awkward name and should be changed to something like `ub_layer_id` or `ub_layer_label`.

**Comment by ShriyaPalsamudram** [comment]
Agreed, removed the logging commit from this PR

**Review by ptrendx — APPROVED** [review]

**Review by timmoon10 — APPROVED** [review]
LGTM

**Comment by ptrendx** [comment]
@ShriyaPalsamudram please sign your commit.

**Review by ksivaman — APPROVED** [review]

**Pull request merged**